### PR TITLE
Avoid losing waitable handles while using MultiThreadedExecutor

### PR DESCRIPTION
**Pull request by Barry-Xu-2018**
Address #2012

This patch may not be the best fix.   
But I hope to find the best fix by discussing this patch.

### CODE REVIEW

File: rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp [hunk 1]
```diff
@@ -120,8 +120,8 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
       }
     }
     for (size_t i = 0; i < waitable_handles_.size(); ++i) {
-      if (!waitable_handles_[i]->is_ready(wait_set)) {
-        waitable_handles_[i].reset();
+      if (waitable_handles_[i]->is_ready(wait_set)) {
+        waitable_triggered_handles_.emplace_back(waitable_handles_[i]);
```
Comment (iuhilnehc-ynos) [inline rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp:124]:
```suggestion
        waitable_triggered_handles_.emplace_back(std::move(waitable_handles_[i]));
```

and add `#include <utility>` before.

File: rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp [hunk 2]
```diff
@@ -392,16 +389,18 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
     rclcpp::AnyExecutable & any_exec,
     const WeakCallbackGroupsToNodesMap & weak_groups_to_nodes) override
   {
-    auto it = waitable_handles_.begin();
-    while (it != waitable_handles_.end()) {
+    auto & waitable_handles =
```
Comment (iuhilnehc-ynos) [inline rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp:392]: This seems to be a trick to make the `TEST_F(TestAllocatorMemoryStrategy, get_next_waitable)` passed.

**Review by iuhilnehc-ynos — APPROVED** [review]
Personally, I can't figure out a better way to fix the issue.
So it LGTM.

File: rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp (hunk 2 above)
Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp:392]: Yes. Maybe need to re-write get_next_waitable.

**Review by fujitatomoya — APPROVED** [review]
lgtm

**Comment by fujitatomoya** [comment]
This changes ABI, so that we cannot backport this to humble. but i believe this is the problem must be addressed to Iron Release. I will start the CI.

@clalancette @wjwwood requesting another review on this.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18122/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12653/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18773/)

**Comment by fujitatomoya** [comment]
@Barry-Xu-2018 can you rebase this? we are having build failure, see https://ci.ros2.org/job/ci_linux/18122/console

**Comment by Barry-Xu-2018** [comment]
Rpr__rclcpp__ubuntu_jammy_amd64 failed since latest code use service_introspection head file which isn't included in build system.

**Comment by Barry-Xu-2018** [comment]
@fujitatomoya Rebase was done.

File: rclcpp/test/rclcpp/strategies/test_allocator_memory_strategy.cpp [hunk 3]
```diff
@@ -657,20 +708,129 @@ TEST_F(TestAllocatorMemoryStrategy, get_next_timer) {
...
-  node2->get_node_waitables_interface()->add_waitable(waitable2, nullptr);
-
   auto get_next_entity = [this](const WeakCallbackGroupsToNodesMap & weak_groups_to_nodes) {
       rclcpp::AnyExecutable result;
       allocator_memory_strategy()->get_next_waitable(result, weak_groups_to_nodes);
       return result;
     };
 
-  EXPECT_TRUE(TestGetNextEntity(node1, node2, get_next_entity));
+  {
```
Comment (iuhilnehc-ynos) [inline rclcpp/test/rclcpp/strategies/test_allocator_memory_strategy.cpp:717]: I think all cases in this file(that's why it is located in `test/strategy` directory) are intended to test the basic interfaces of `AllocatorMemoryStrategy`. I am afraid that we might be not allowed to use the `rcl_wait` in the test here.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18138/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12674/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18786/)

**Comment by Barry-Xu-2018** [comment]
After checking,   
Linux failure is related to `python_qt_binding`. So it isn't related to this patch.   
Linux-aarch64 failure is related to internal error for CI system.

**Comment by clalancette** [comment]
> Linux-aarch64 failure is related to internal error for CI system.

Yeah, it got re-run.  I've updated Tomoya's post to show the re-run.

While I don't think this PR is causing the CI to go yellow, I'm going to re-run this anyway:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18154/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12691/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18812/)

**Review by mjcarroll — APPROVED** [review]

**Comment by fujitatomoya** [comment]
this cannot be backport to humble since it requires ABI change.

**Comment by wjwwood** [comment]
Just to help with cross-referencing, we're discussing in a few places whether or not this is the right fix for this issue: https://github.com/ros2/rclcpp/pull/2371#discussion_r1441172747

I'm currently leaning towards the idea that this is not the right fix and instead we should change specific waitables to address this issue.

**Pull request merged**